include/alpaka/mem/view/Traits.hpp

```diff
@@ -250,9 +250,11 @@ namespace alpaka
 
             //-----------------------------------------------------------------------------
             //! \return The pitch in bytes. This is the distance in bytes between two consecutive elements in the given dimension.
+            ALPAKA_NO_HOST_ACC_WARNING
```

Comment:
I removed all of this stuff in #796, why do you reintroduce it? Since all of this code is purely CPU code there is no need for any annotation.

Reply:
ah ok, I was already wondering.
It is **not** purely CPU code and therefore HCC does not like that changes and fails with errors like:
> /alpaka/include/alpaka/vec/Vec.hpp:198:22: error:  'alpaka::mem::view::detail::CreatePitchBytes<0>::create':  no overloaded function has restriction specifiers that are compatible with the ambient context 'createVecFromIndexedFnArbitrary'

Let's take this as an example to see where it fails exactly.
(_Right after writing this post I am going to check for the HIP-clang to see whether it is less restrictive, because HCC still behaves like C++AMP, although support for it has been removed as of ROCm 1.9. HCC restrictions also tend to ripple through the call hierarchy like "restrict" from C++AMP. If HIP-clang works, then it will be ToDo bringing it into travis/CI._)

Ok, that function chain starts with `alpaka::meta::forEachType()` (which is already flagged as host-device).
Let's check the callstack with that entry function forEachType (in `/alpaka/test/unit/mem/view/src//ViewPlainPtrTest.cpp:223`):
```c++
alpaka::meta::forEachType< alpaka::test::acc::TestAccs >( TestTemplatePlain() ); /* unspecified, entry function, /alpaka/test/unit/mem/view/src//ViewPlainPtrTest.cpp:223 */
forEachTypeHelper() /* host-device */
f.template operator() /* unspecified, see ViewPlainPtrTest.cpp:195 */
alpaka::test::mem::view::testViewPlainPtr() /* unspecified, see ViewPlainPtrTest.cpp:96 */
alpaka::mem::view::getPitchBytesVec /* unspecified, see mem/view/Traits.hpp:550 */
vec::createVecFromIndexedFnWorkaround /* host-device, see Vec.hpp:512 */
createVecFromIndexedFn /* host-device, see Vec.hpp:207 */
createVecFromIndexedFnArbitrary /* host-device, see Vec.hpp:190 */
CreatePitchBytes<0>::create /* this must be host-device, see below */
```

The `getPitchBytesVec()` invokes the host-device function `createVecFromIndexedFnArbitrary` (within `vec::createVecFromIndexedFnWorkaround`), which directly invokes `CreatePitchBytes::create()`. 
When `CreatePitchBytes::create()` is unspecified, HCC fails because of unmatched ambient context information.

```c++
return
  vec::createVecFromIndexedFnWorkaround< /* getPitchBytesVec() in there */
    dim::Dim<TPitch>,                                                                                                                                                                      
    idx::Idx<TPitch>,                                                                                                                                                                      
    detail::CreatePitchBytes /* its create() method must match context of getPitchBytesVec() */
  >(pitch);
```

Full backtrace here:

<details>

```
In file included from /alpaka/test/unit/mem/view/src//ViewPlainPtrTest.cpp:11:
In file included from /alpaka/include/alpaka/alpaka.hpp:30:
In file included from /alpaka/include/alpaka/acc/AccGpuHipRt.hpp:22:
In file included from /alpaka/include/alpaka/workdiv/WorkDivHipBuiltIn.hpp:21:
In file included from /alpaka/include/alpaka/workdiv/Traits.hpp:17:
/alpaka/include/alpaka/vec/Vec.hpp:198:22: error:  'alpaka::mem::view::detail::CreatePitchBytes<0>::create':  no overloaded function has restriction specifiers that are compatible with the ambient context 'createVecFromIndexedFnArbitrary'
                    (TTFnObj<TIndices>::create(std::forward<TArgs>(args)...))...);
                     ^
/alpaka/include/alpaka/vec/Vec.hpp:212:21: note: in instantiation of function template specialization 'alpaka::vec::Vec<std::integral_constant<unsigned long, 1>, long>::createVecFromIndexedFnArbitrary<CreatePitchBytes, const alpaka::mem::buf::BufHipRt<float, std::integral_constant<unsigned long, 1>, long> &, unsigned long, 0>' requested here
                    createVecFromIndexedFnArbitrary<
                    ^
/alpaka/include/alpaka/vec/Vec.hpp:523:17: note: in instantiation of function template specialization 'alpaka::vec::Vec<std::integral_constant<unsigned long, 1>, long>::createVecFromIndexedFn<CreatePitchBytes, const alpaka::mem::buf::BufHipRt<float, std::integral_constant<unsigned long, 1>, long> &>' requested here
                createVecFromIndexedFn<
                ^
/alpaka/include/alpaka/mem/view/Traits.hpp:555:26: note: in instantiation of function template specialization 'alpaka::vec::createVecFromIndexedFnWorkaround<std::integral_constant<unsigned long, 1>, long, detail::CreatePitchBytes, const alpaka::mem::buf::BufHipRt<float, std::integral_constant<unsigned long, 1>, long> &>' requested here
                    vec::createVecFromIndexedFnWorkaround<
                         ^
/alpaka/test/unit/mem/view/src//ViewPlainPtrTest.cpp:117:32: note: in instantiation of function template specialization 'alpaka::mem::view::getPitchBytesVec<alpaka::mem::buf::BufHipRt<float, std::integral_constant<unsigned long, 1>, long> >' requested here
            alpaka::mem::view::getPitchBytesVec(buf));
                               ^
/alpaka/test/unit/mem/view/src//ViewPlainPtrTest.cpp:197:34: note: in instantiation of function template specialization 'alpaka::test::mem::view::testViewPlainPtr<alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 1>, long>, float>' requested here
        alpaka::test::mem::view::testViewPlainPtr<TAcc, float>();
                                 ^
/alpaka/include/alpaka/meta/ForEachType.hpp:72:32: note: in instantiation of function template specialization 'TestTemplatePlain::operator()<alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 1>, long> >' requested here
                    f.template operator()<T>(
                               ^
/alpaka/include/alpaka/meta/ForEachType.hpp:99:15: note: in instantiation of function template specialization 'alpaka::meta::detail::ForEachTypeHelper<std::tuple<alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 1>, long>, alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 2>, long>, alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 1>, unsigned long>, alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 2>, unsigned long>, alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 1>, int>, alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 2>, int>, alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 1>, unsigned int>, alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 2>, unsigned int>, alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 1>, short>, alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 2>, short>, alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 1>, unsigned short>, alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 2>, unsigned short> > >::forEachTypeHelper<TestTemplatePlain>' requested here
            ::forEachTypeHelper(
              ^
/alpaka/test/unit/mem/view/src//ViewPlainPtrTest.cpp:223:19: note: in instantiation of function template specialization 'alpaka::meta::forEachType<std::tuple<alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 1>, long>, alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 2>, long>, alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 1>, unsigned long>, alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 2>, unsigned long>, alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 1>, int>, alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 2>, int>, alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 1>, unsigned int>, alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 2>, unsigned int>, alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 1>, short>, alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 2>, short>, alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 1>, unsigned short>, alpaka::acc::AccGpuHipRt<std::integral_constant<unsigned long, 2>, unsigned short> >, TestTemplatePlain>' requested here
    alpaka::meta::forEachType< alpaka::test::acc::TestAccs >( TestTemplatePlain() );
```

</details>

Reply:
update: trying to figure out hip-clang build workflow, [also reporting status on /HIP](https://github.com/ROCm-Developer-Tools/HIP/issues/1196)

Reply:
we analyzed this behavior with @tdd11235813  and found out that `hipcc` was not providing error via the return code. Therefore I do not see that some tests in #796 failed. The `hipcc` bug was already fixed in https://github.com/ROCm-Developer-Tools/HIP/pull/1117 but are for some reason not part of the last releases.